tailnet/controllers.go

```diff
@@ -0,0 +1,361 @@
+}
+
+type ResumeTokenClient interface {
+	RefreshResumeToken(ctx context.Context, in *proto.RefreshResumeTokenRequest) (*proto.RefreshResumeTokenResponse, error)
+}
+
+type ResumeTokenController interface {
+	New(ResumeTokenClient) CloserWaiter
+	Token() (string, bool)
+}
+
+type TelemetryClient interface {
+	PostTelemetry(ctx context.Context, in *proto.TelemetryRequest) (*proto.TelemetryResponse, error)
+}
+
+type TelemetryController interface {
+	New(TelemetryClient)
+}
+
+// ControlProtocolClients represents an abstract interface to the tailnet control plane via a set
+// of protocol clients. The Closer should close all the clients (e.g. by closing the underlying
+// connection).
+type ControlProtocolClients struct {
+	Closer      io.Closer
+	Coordinator CoordinatorClient
+	DERP        DERPClient
+	ResumeToken ResumeTokenClient
+	Telemetry   TelemetryClient
+}
+
+type ControlProtocolDialer interface {
+	// Dial connects to the tailnet control plane and returns clients for the different control
+	// sub-protocols (coordination, DERP maps, resume tokens, and telemetry).  If the
+	// ResumeTokenController is not nil, the dialer should query for a resume token and use it to
+	// dial, if available.
+	Dial(ctx context.Context, r ResumeTokenController) (ControlProtocolClients, error)
+}
+
+// basicCoordinationController handles the basic coordination operations common to all types of
+// tailnet consumers:
+//
+//  1. sending local node updates to the Coordinator
+//  2. receiving peer node updates and programming them into the Coordinatee (e.g. tailnet.Conn)
+//  3. (optionally) sending ReadyToHandshake acknowledgements for peer updates.
+type basicCoordinationController struct {
+	logger      slog.Logger
+	coordinatee Coordinatee
+	sendAcks    bool
+}
+
+func (c *basicCoordinationController) New(client CoordinatorClient) CloserWaiter {
+	b := &basicCoordination{
+		logger:       c.logger,
+		errChan:      make(chan error, 1),
+		coordinatee:  c.coordinatee,
+		client:       client,
+		respLoopDone: make(chan struct{}),
+		sendAcks:     c.sendAcks,
+	}
+
+	c.coordinatee.SetNodeCallback(func(node *Node) {
+		pn, err := NodeToProto(node)
+		if err != nil {
+			b.logger.Critical(context.Background(), "failed to convert node", slog.Error(err))
+			b.sendErr(err)
+			return
+		}
+		b.Lock()
+		defer b.Unlock()
+		if b.closed {
+			b.logger.Debug(context.Background(), "ignored node update because coordination is closed")
+			return
+		}
+		err = b.client.Send(&proto.CoordinateRequest{UpdateSelf: &proto.CoordinateRequest_UpdateSelf{Node: pn}})
+		if err != nil {
+			b.sendErr(xerrors.Errorf("write: %w", err))
+		}
+	})
+	go b.respLoop()
+
+	return b
+}
+
+type basicCoordination struct {
+	sync.Mutex
+	closed       bool
+	errChan      chan error
+	coordinatee  Coordinatee
+	logger       slog.Logger
+	client       CoordinatorClient
+	respLoopDone chan struct{}
+	sendAcks     bool
+}
+
+func (c *basicCoordination) Close(ctx context.Context) (retErr error) {
```

Comment:
This context is only ever used to log a warning if it expires - although this is in the original code, it looks a bit awkward. It's more noticeable (and potentially more annoying) now that it's on the interface `CloserWaiter`. Do we need it?

Reply:
It's not just logging; it's how long we wait for the remote end to hang up before we forcefully close our end.

Reply:
🤦 my bad